### PR TITLE
feat(api): add external API endpoint for creating urls

### PR DESCRIPTION
**Pull request by halfwhole**
## Problem

We want to add an external API endpoint for creating URLs (`POST /api/v1/urls`)

See [Notion page on external API](https://www.notion.so/opengov/20220922-Go-External-API-e3557aa9bf64435c89d9e9ff8c1c638a)

## Solution

**Features**:
- Created a new `ApiV1Controller` in `src/server/modules/api/external-v1`, with the `createUrl` method
- Request fields to create URL: `shortUrl` (optional), `longUrl`
- Response fields for created URL (as defined by `UrlV1DTO` type): `shortUrl`, `longUrl`, `state`, `clicks`, `createdAt`, `updatedAt` - note that certain fields e.g. `tags`, `description` have been restricted
  - ~To map from the original DTO `StorableUrl` to `UrlV1DTO`, I wrote a private function `sanitizeUrlForApi` in the controller that picks out only the fields that we want to return - is this okay, or is there a better method? Will be expecting this function to be reused in the future for other parts of the API controller~
  - ~Decided against creating a new mapper, because mappers seem specific to transforming from persistent -> DTO~
  - EDIT: we'll create a new mapper `UrlV1Mapper` to map from `StorableUrl` to `UrlV1DTO`, which should do for now
- `UrlManagementService` and `UrlRepository` have been changed to additionally take in a `source` field, to distinguish between console vs. API created links
- Added a fallback 404 for all non-matching `/api/v1/<...>` routes

**Improvements**:

- Made a number of small edits to the error messages, this may clutter the files changed (sorry!)
- The backend currently returns the error message `Short link "<shortUrl>" is used. Click here to find out more` when the short URL is already taken. But it seems weird to return "Click here to find out more" from the backend - the external API definitely shouldn't have that, so I've removed it from the external API but preserved it in the internal API. In the future, we should shift the "Click here" part to the frontend

**Future possible improvements**: 
- Change _internal_ API endpoint from `/api/user/url` to `/api/user/urls`, consistent with the external API `/api/v1/urls`
- Restrict user from submitting `userId` in request body when creating URL through external API

## Tests

- [x] Unit tests for API controller - these borrow heavily from the unit tests for the original User controller

Should write integration tests too, when we have them 

## Deploy notes

New environment variables:
- (Optional) `API_LINK_RANDOM_STR_LENGTH`: length of randomly generated `shortUrl` for API created links, defaults to 8

### CODE REVIEW

File: src/server/modules/api/external-v1/ApiController.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import { UrlManagementService } from '../../user/interfaces'
+import { MessageType } from '../../../../shared/util/messages'
+import { StorableUrlSource } from '../../../repositories/enums'
+import { StorableUrl } from '../../../repositories/types'
+
+import { ApiUrl, UrlCreationRequest } from '.'
+
+@injectable()
+export class ApiController {
```
Comment (thanhdatle) [inline src/server/modules/api/external-v1/ApiController.ts:18]: I think we can name this ApiV1Controller

File: src/server/modules/api/external-v1/index.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+type ShortUrlProperty = {
+  shortUrl: string
+}
+
+type ShortUrlOperationProperty = UserIdProperty & ShortUrlProperty
+
+export type UrlCreationRequest = ShortUrlOperationProperty &
+  OptionalLongUrlProperty
+
+export type ApiUrl = Pick<
```
Comment (thanhdatle) [inline src/server/modules/api/external-v1/index.ts:21]: May be UrlV1DTO, 🤔

File: src/server/modules/user/UserController.ts [hunk 3]
```diff
@@ -91,14 +92,15 @@ export class UserController {
         return
       }
       if (error instanceof AlreadyExistsError) {
-        res.badRequest(jsonMessage(error.message, MessageType.ShortUrlError))
+        const newErrorMessage = `${error.message} Click here to find out more` // TODO: shift message to frontend
```
Comment (thanhdatle) [inline src/server/modules/user/UserController.ts:95]: Please help to create a Notion task instead

**Review by thanhdatle — CHANGES_REQUESTED** [review]

File: src/server/modules/api/external-v1/ApiController.ts (hunk 1 above)
Comment (halfwhole) [reply src/server/modules/api/external-v1/ApiController.ts]: done in [ad18d7c](https://github.com/opengovsg/GoGovSG/pull/2069/commits/ad18d7c82fee948a7aafe8a1107a36870589304f)

File: src/server/modules/api/external-v1/index.ts (hunk 2 above)
Comment (halfwhole) [reply src/server/modules/api/external-v1/index.ts]: done in [07bce66](https://github.com/opengovsg/GoGovSG/pull/2069/commits/07bce665fac53138bbc48a6b793465c409637ecf)

File: src/server/modules/user/UserController.ts (hunk 3 above)
Comment (halfwhole) [reply src/server/modules/user/UserController.ts]: removed TODO comment in [c2e319e](https://github.com/opengovsg/GoGovSG/pull/2069/commits/c2e319e3f149624b992d8f31d463426887dce30b), have created notion task

File: src/server/modules/api/external-v1/ApiController.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      if (error instanceof Sequelize.ValidationError) {
+        res.badRequest(jsonMessage(error.message))
+      }
+      logger.error(`Error creating short URL:\t${error}`)
+      res.badRequest(jsonMessage('Server error.'))
+      return
+    }
+  }
+
+  private sanitizeUrlForApi: (url: StorableUrl) => ApiUrl = (url) => {
```
[the remark this answers is not in the document]
Comment (halfwhole) [reply src/server/modules/api/external-v1/ApiController.ts]: done in [24e4ba6](https://github.com/opengovsg/GoGovSG/pull/2069/commits/24e4ba6fb19c7667c5826516f7fb1df09cc52de0)

**Comment by halfwhole** [comment]
^ ignore the above gitguardian warning, it's from the previously-defined `API_KEY_SALT` in `docker-compose.yml` which was left untouched in this PR

Realise that `shortUrl` should be optional instead of required for API-created URLs, added it in https://github.com/opengovsg/GoGovSG/pull/2069/commits/a5815c6ebd752b7d4bef8126b430468e4365dd12. The `shortUrl` has a randomly generated length according to the optional env variable `API_LINK_RANDOM_STR_LENGTH`, defaults to 8 like bulk created links

**Review by thanhdatle — APPROVED** [review]

**Pull request merged**